From 53508591c67774cffefcb0d607c297ea87340817 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 25 May 2024 20:33:39 +0200
Subject: [PATCH 04/14] Change Scheduler BukkitQuestsTopCommand

---
 .../commands/quests/subcommands/BukkitQuestsTopCommand.java     | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/core/src/main/java/me/pikamug/quests/commands/quests/subcommands/BukkitQuestsTopCommand.java b/core/src/main/java/me/pikamug/quests/commands/quests/subcommands/BukkitQuestsTopCommand.java
index 3fda1473c..77bac4ca6 100644
--- a/core/src/main/java/me/pikamug/quests/commands/quests/subcommands/BukkitQuestsTopCommand.java
+++ b/core/src/main/java/me/pikamug/quests/commands/quests/subcommands/BukkitQuestsTopCommand.java
@@ -83,7 +83,7 @@ public class BukkitQuestsTopCommand extends BukkitQuestsSubCommand {
                         .replace("<greatest>", String.valueOf(plugin.getConfigSettings().getTopLimit())));
                 return;
             }
-            Bukkit.getScheduler().runTaskAsynchronously(plugin, () -> {
+            Bukkit.getAsyncScheduler().runNow(plugin, (task) -> {
                 final File folder = new File(plugin.getDataFolder(), "data");
                 final File[] playerFiles = folder.listFiles();
                 final Map<String, Integer> questPoints = new HashMap<>();
-- 
2.45.1.windows.1

