From a4dc99a43aa8edce0fb2228f6b1b38e11de801ac Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sat, 25 May 2024 20:33:41 +0200
Subject: [PATCH 05/14] Change Scheduler BukkitDependencies

---
 .../quests/dependencies/BukkitDependencies.java       | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/core/src/main/java/me/pikamug/quests/dependencies/BukkitDependencies.java b/core/src/main/java/me/pikamug/quests/dependencies/BukkitDependencies.java
index 8a58f8e02..6acc3f488 100644
--- a/core/src/main/java/me/pikamug/quests/dependencies/BukkitDependencies.java
+++ b/core/src/main/java/me/pikamug/quests/dependencies/BukkitDependencies.java
@@ -20,6 +20,7 @@ import com.herocraftonline.heroes.Heroes;
 import com.herocraftonline.heroes.characters.Hero;
 import com.herocraftonline.heroes.characters.classes.HeroClass;
 import io.github.znetworkw.znpcservers.npc.NPC;
+import io.papermc.paper.threadedregions.scheduler.ScheduledTask;
 import lol.pyr.znpcsplus.ZNPCsPlus;
 import lol.pyr.znpcsplus.api.NpcApi;
 import lol.pyr.znpcsplus.api.NpcApiProvider;
@@ -67,7 +68,7 @@ public class BukkitDependencies implements Dependencies {
     public static ZNPCsPlus znpcsPlusLegacy = null;
     public static NpcApi znpcsPlusApi = null;
     private static PartiesAPI parties = null;
-    private int npcEffectThread = -1;
+    private ScheduledTask npcEffectThread = null;
     
     public BukkitDependencies(final BukkitQuestsPlugin plugin) {
         this.plugin = plugin;
@@ -381,15 +382,15 @@ public class BukkitDependencies implements Dependencies {
     }
 
     public void startNpcEffectThread() {
-        if (npcEffectThread == -1 && plugin.getConfigSettings().canNpcEffects()) {
-            npcEffectThread = plugin.getServer().getScheduler().scheduleSyncRepeatingTask(plugin,
-                    plugin.getNpcEffectThread(), 20, 20);
+        if (npcEffectThread == null && plugin.getConfigSettings().canNpcEffects()) {
+            npcEffectThread = plugin.getServer().getGlobalRegionScheduler().runAtFixedRate(plugin,
+                    task -> plugin.getNpcEffectThread().run(), 20, 20);
         }
     }
 
     @SuppressWarnings("unused")
     public void stopNpcEffectThread() {
-        plugin.getServer().getScheduler().cancelTask(npcEffectThread);
+        npcEffectThread.cancel();
     }
 
     public SkillType getMcMMOSkill(final String s) {
-- 
2.45.1.windows.1

