From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sun, 5 Jul 2026 21:32:40 +0200
Subject: [PATCH] Add Build Gradle


diff --git a/.gitignore b/.gitignore
index 5ff6309b7199129c1afe4f4ec1906e640bec48c6..8a3ce194168e38d898a7bba9a48ee06a1218ea27 100644
--- a/.gitignore
+++ b/.gitignore
@@ -35,4 +35,8 @@ build/
 .vscode/
 
 ### Mac OS ###
-.DS_Store
\ No newline at end of file
+.DS_Store
+
+### Gradle ###
+.gradle/
+!gradle/wrapper/gradle-wrapper.jar
\ No newline at end of file
diff --git a/build.gradle.kts b/build.gradle.kts
new file mode 100644
index 0000000000000000000000000000000000000000..19c1a61ae7e3d8db672cbba89af32e691b001756
--- /dev/null
+++ b/build.gradle.kts
@@ -0,0 +1,33 @@
+plugins {
+    `java-library`
+}
+
+group = "me.pikamug.localelib"
+version = "4.1.4"
+
+java {
+    sourceCompatibility = JavaVersion.VERSION_21
+    targetCompatibility = JavaVersion.VERSION_21
+}
+
+repositories {
+    mavenCentral()
+    maven("https://repo.papermc.io/repository/maven-public/")
+}
+
+dependencies {
+    compileOnly("io.papermc.paper:paper-api:1.20.6-R0.1-SNAPSHOT")
+}
+
+tasks.withType<JavaCompile>().configureEach {
+    options.encoding = "UTF-8"
+}
+
+tasks.processResources {
+    filteringCharset = "UTF-8"
+    val props = mapOf("version" to project.version.toString())
+    inputs.properties(props)
+    filesMatching("plugin.yml") {
+        expand("project" to props)
+    }
+}
\ No newline at end of file
diff --git a/settings.gradle.kts b/settings.gradle.kts
new file mode 100644
index 0000000000000000000000000000000000000000..7dfb6e97edb6411ae06630fe6b9c2eb938145892
--- /dev/null
+++ b/settings.gradle.kts
@@ -0,0 +1 @@
+rootProject.name = "LocaleLib"
\ No newline at end of file
