From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sun, 5 Jul 2026 21:34:12 +0200
Subject: [PATCH] Fix support Potion


diff --git a/src/main/java/me/pikamug/localelib/LocaleManager.java b/src/main/java/me/pikamug/localelib/LocaleManager.java
index 965687c47363aba5139a5fc1ef7bec6f81f272fd..da38d7b08ba43fab8c7270146689f5e56c5224a3 100644
--- a/src/main/java/me/pikamug/localelib/LocaleManager.java
+++ b/src/main/java/me/pikamug/localelib/LocaleManager.java
@@ -35,7 +35,7 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.EnchantmentStorageMeta;
 import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.inventory.meta.PotionMeta;
-import org.bukkit.potion.Potion;
+import org.bukkit.potion.PotionType;
 
 import java.io.IOException;
 import java.lang.reflect.Method;
@@ -119,20 +119,26 @@ public class LocaleManager{
         if (player == null || itemStack == null) {
             return false;
         }
-        if (!hasBasePotionData && itemStack.getType().equals(Material.POTION)) {
-            final Potion potion = Potion.fromItemStack(itemStack);
+        final Material type = itemStack.getType();
+        if (type == Material.POTION || type == Material.SPLASH_POTION || type == Material.LINGERING_POTION) {
             String msg = message;
-            if (potion.isSplash()) {
-                msg = message.replace("<item>", "<prefix> <item>")
+
+            if (type == Material.SPLASH_POTION) {
+                msg = msg.replace("<item>", "<prefix> <item>")
                         .replace("<prefix>", translate(message, "potion.prefix.grenade", "<prefix>"));
             }
             String potionName = "item.potion.name";
-            if (potion.getType().getEffectType() != null) {
-                potionName = oldPotions1dot8.get(potion.getType().getEffectType().getName());
+            if (itemStack.getItemMeta() instanceof PotionMeta potionMeta) {
+                PotionType basePotionType = potionMeta.getBasePotionType(); // API 1.20.2+
+                if (basePotionType != null && !basePotionType.getPotionEffects().isEmpty()) {
+                    String effectName = basePotionType.getPotionEffects().getFirst().getType().getName();
+                    potionName = oldPotions1dot8.getOrDefault(effectName, potionName);
+                }
             }
-            msg = msg.replace("<item>", translate(message, potionName, "<item>"));
 
-            Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + msg + "\"]");
+            msg = msg.replace("<item>", translate(message, potionName, "<item>"));
+            Bukkit.dispatchCommand(Bukkit.getConsoleSender(),
+                    "tellraw " + formatName(player) + " [\"" + msg + "\"]");
             return true;
         }
         return sendMessage(player, message, itemStack.getType(), itemStack.getDurability(), itemStack.getEnchantments(),
@@ -203,7 +209,7 @@ public class LocaleManager{
      * @param enchantments Enchantments for the item being translated
      */
     public boolean sendMessage(final Player player, final String message, final Material material, final short durability,
-            final Map<Enchantment, Integer> enchantments) {
+                               final Map<Enchantment, Integer> enchantments) {
         return sendMessage(player, message, material, durability, enchantments, null);
     }
 
