From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Euphyllia Bierque <bierque.euphyllia@gmail.com>
Date: Sun, 5 Jul 2026 21:39:35 +0200
Subject: [PATCH] Dispatch command on global thread


diff --git a/src/main/java/me/pikamug/localelib/LocaleManager.java b/src/main/java/me/pikamug/localelib/LocaleManager.java
index da38d7b08ba43fab8c7270146689f5e56c5224a3..bb80336f20747f05025a9c236f573b9c4eef2f33 100644
--- a/src/main/java/me/pikamug/localelib/LocaleManager.java
+++ b/src/main/java/me/pikamug/localelib/LocaleManager.java
@@ -35,6 +35,7 @@ import org.bukkit.inventory.ItemStack;
 import org.bukkit.inventory.meta.EnchantmentStorageMeta;
 import org.bukkit.inventory.meta.ItemMeta;
 import org.bukkit.inventory.meta.PotionMeta;
+import org.bukkit.plugin.Plugin;
 import org.bukkit.potion.PotionType;
 
 import java.io.IOException;
@@ -62,6 +63,7 @@ public class LocaleManager{
     private final Map<String, String> oldSplashPotions = LocaleKeys.getSplashPotionKeys();
     private final Map<String, String> oldEntities = LocaleKeys.getEntityKeys();
     private Map<String, String> englishTranslations;
+    private Plugin plugin;
 
     public LocaleManager() {
         oldVersion = isBelow113();
@@ -103,6 +105,10 @@ public class LocaleManager{
         }
     }
 
+    public void setPlugin(Plugin plugin) {
+        this.plugin = plugin;
+    }
+
     /**
      * Send message with item name translated to the client's locale.
      * ItemStack is required. This method supports 1.8 potions.<p>
@@ -137,8 +143,9 @@ public class LocaleManager{
             }
 
             msg = msg.replace("<item>", translate(message, potionName, "<item>"));
-            Bukkit.dispatchCommand(Bukkit.getConsoleSender(),
-                    "tellraw " + formatName(player) + " [\"" + msg + "\"]");
+            final String messageToSend = msg;
+            Bukkit.getGlobalRegionScheduler().run(plugin,scheduledTask -> Bukkit.dispatchCommand(Bukkit.getConsoleSender(),
+                    "tellraw " + formatName(player) + " [\"" + messageToSend + "\"]"));
             return true;
         }
         return sendMessage(player, message, itemStack.getType(), itemStack.getDurability(), itemStack.getEnchantments(),
@@ -188,7 +195,9 @@ public class LocaleManager{
         for (final String lk : lvlKeys) {
             msg = msg.replaceFirst("<level>",  translate(msg, lk, "<level>"));
         }
-        Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + msg + "\"]");
+        final String messageToSend = msg;
+        Bukkit.getGlobalRegionScheduler().run(plugin,scheduledTask ->
+                Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + messageToSend + "\"]"));
         return true;
     }
 
@@ -240,7 +249,9 @@ public class LocaleManager{
                 msg = message.replaceFirst("<level>",  translate(message, lk, "<level>"));
             }
         }
-        Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + msg + "\"]");
+        final String messageToSend = msg;
+        Bukkit.getGlobalRegionScheduler().run(plugin,scheduledTask ->
+                Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + messageToSend + "\"]"));
         return true;
     }
 
@@ -262,7 +273,9 @@ public class LocaleManager{
         }
         final String key = queryEntityType(type, extra);
         final String msg = message.replace("<mob>", translate(message, key, "<mob>"));
-        Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + msg + "\"]");
+        final String messageToSend = msg;
+        Bukkit.getGlobalRegionScheduler().run(plugin, scheduledTask ->
+                Bukkit.dispatchCommand(Bukkit.getConsoleSender(), "tellraw " + formatName(player) + " [\"" + messageToSend + "\"]"));
         return true;
     }
 
